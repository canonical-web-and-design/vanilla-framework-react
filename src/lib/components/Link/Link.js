import React from 'react';
import PropTypes from 'prop-types';
import './Link.scss';

const Link = (props) => {
  let classArray = [];

  if (props.soft) {
    classArray = [...classArray, 'p-link--soft'];
  }

  if (props.strong) {
    classArray = [...classArray, 'p-link--strong'];
  }

  if (props.inverted) {
    classArray = [...classArray, 'p-link--inverted'];
  }

  if (props.external) {
    classArray = [...classArray, 'p-link--external'];
  }

  const classString = classArray.join(' ');

  if (props.top) {
    return (
      <div className="p-top">
        <a href={props.href} className={`${classString} p-top__link`}>{props.children}</a>
      </div>
    );
  }

  return (
    <a href={props.href} className={classString}>{props.children}</a>
  );
};

Link.defaultProps = {
  soft: false,
  strong: false,
  inverted: false,
  external: false,
  top: false,
};

Link.propTypes = {
  children: PropTypes.node.isRequired,
  href: PropTypes.string.isRequired,
  soft: PropTypes.bool,
  strong: PropTypes.bool,
  inverted: PropTypes.bool,
  external: PropTypes.bool,
  top: PropTypes.bool,
};

Link.displayName = 'Link';

export default Link;
