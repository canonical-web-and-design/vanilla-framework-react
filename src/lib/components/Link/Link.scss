@import 'vanilla-framework/scss/base';
@import 'vanilla-framework/scss/patterns_links';
@include vf-p-links;
