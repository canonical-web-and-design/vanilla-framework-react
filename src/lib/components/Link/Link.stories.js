import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';

import Link from './Link';

storiesOf('Link', module)
  .add('Default',
    withInfo('The default Link component.')(() => (
      <Link href="https://vanilla-framework.github.io/vanilla-framework-react/">Default Link</Link>),
    ),
  )

  .add('External',
    withInfo('The "external" modifier should be used on Link components that go to a different domain than the current one. ')(() => (
      <Link external href="https://vanillaframework.io/">External Link</Link>),
    ),
  )

  .add('Soft',
    withInfo('The "soft" modifier should be used on Link components where many links are grouped together, such as a link cloud.')(() => (
      <Link soft href="https://vanilla-framework.github.io/vanilla-framework-react/">Soft Link</Link>),
    ),
  )

  .add('Strong',
    withInfo('The "strong" modifier should be used on Link components that require emphasis or on a dark background.')(() => (
      <Link strong href="https://vanilla-framework.github.io/vanilla-framework-react/">Strong Link</Link>),
    ),
  )

  .add('Inverted',
    withInfo('The "inverted" modifier should be used where Link components are placed on a dark background.')(() => (
      <div style={{ backgroundColor: '#333' }}>
        <Link inverted href="https://vanilla-framework.github.io/vanilla-framework-react/">Inverted Link</Link>
      </div>),
    ),
  )

  .add('Back to top',
    withInfo('The "top" prop can be used to make it easier to go back to the top on long pages. If the page is divided into different sections, you can use more than one per page.')(() => (
      <Link top href="https://vanilla-framework.github.io/vanilla-framework-react/">Back to top</Link>),
    ),
  )

  .add('Combination',
    withInfo('You can combine Link modifiers.')(() => (
      <div>
        <p><Link external strong href="https://vanillaframework.io/">External/Strong Link</Link></p>
        <p><Link external soft href="https://vanillaframework.io/">External/Soft Link</Link></p>
        <p><Link top strong href="https://vanilla-framework.github.io/vanilla-framework-react/">Strong/Back to top Link</Link></p>
      </div>),
    ),
  );
