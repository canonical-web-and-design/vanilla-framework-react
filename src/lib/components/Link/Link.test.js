import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import Link from './Link';

describe('Link component', () => {
  it('should render a simple link correctly', () => {
    const link = ReactTestRenderer.create(
      <Link href="https://vanilla-framework.github.io/vanilla-framework-react/">Default Link</Link>,
    );
    const json = link.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should accept modifiers correctly', () => {
    const link = ReactTestRenderer.create(
      <div>
        <Link external href="https://vanillaframework.io/">External Link</Link>
        <Link soft href="https://vanilla-framework.github.io/vanilla-framework-react/">Soft Link</Link>
        <Link strong href="https://vanilla-framework.github.io/vanilla-framework-react/">Strong Link</Link>
        <Link inverted href="https://vanilla-framework.github.io/vanilla-framework-react/">Inverted Link</Link>
        <Link top href="https://vanilla-framework.github.io/vanilla-framework-react/">Back to top</Link>
        <Link external strong href="https://vanillaframework.io/">External/Strong Link</Link>
        <Link external soft href="https://vanillaframework.io/">External/Soft Link</Link>
        <Link top strong href="https://vanilla-framework.github.io/vanilla-framework-react/">Strong/Back to top Link</Link>
      </div>,
    );
    const json = link.toJSON();
    expect(json).toMatchSnapshot();
  });
});
