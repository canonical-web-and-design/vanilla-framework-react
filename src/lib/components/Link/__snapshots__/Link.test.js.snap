// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link component should accept modifiers correctly 1`] = `
<div>
  <a
    className="p-link--external"
    href="https://vanillaframework.io/"
  >
    External Link
  </a>
  <a
    className="p-link--soft"
    href="https://vanilla-framework.github.io/vanilla-framework-react/"
  >
    Soft Link
  </a>
  <a
    className="p-link--strong"
    href="https://vanilla-framework.github.io/vanilla-framework-react/"
  >
    Strong Link
  </a>
  <a
    className="p-link--inverted"
    href="https://vanilla-framework.github.io/vanilla-framework-react/"
  >
    Inverted Link
  </a>
  <div
    className="p-top"
  >
    <a
      className=" p-top__link"
      href="https://vanilla-framework.github.io/vanilla-framework-react/"
    >
      Back to top
    </a>
  </div>
  <a
    className="p-link--strong p-link--external"
    href="https://vanillaframework.io/"
  >
    External/Strong Link
  </a>
  <a
    className="p-link--soft p-link--external"
    href="https://vanillaframework.io/"
  >
    External/Soft Link
  </a>
  <div
    className="p-top"
  >
    <a
      className="p-link--strong p-top__link"
      href="https://vanilla-framework.github.io/vanilla-framework-react/"
    >
      Strong/Back to top Link
    </a>
  </div>
</div>
`;

exports[`Link component should render a simple link correctly 1`] = `
<a
  className=""
  href="https://vanilla-framework.github.io/vanilla-framework-react/"
>
  Default Link
</a>
`;
